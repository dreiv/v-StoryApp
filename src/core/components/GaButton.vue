<script setup lang="ts">
import { computed, useCssModule } from 'vue'

const props = defineProps<{ primary?: boolean; transparent?: boolean; icon?: boolean }>()
const style = useCssModule()

const classes = computed(() => [
  style.base,
  { [style.default]: !props.primary && !props.transparent },
  { [style.primary]: props.primary },
  { [style.transparent]: props.transparent },
  { [style.icon]: props.icon },
])
</script>

<template>
  <button :class="classes"><slot /></button>
</template>

<style module>
.base {
  box-sizing: border-box;
  border-radius: var(--rad-sm);
  padding: 0 var(--sp-md);
  min-width: var(--sp-xxl);
  height: var(--sp-xxl);

  font-weight: 500;
  font-family: Inter, sans-serif;

  &:enabled {
    cursor: pointer;
  }

  &:disabled {
    color: var(--gray-70);
  }
}

.default {
  --border-col: var(--primary-80);

  border: 1px solid var(--border-col);
  background-color: var(--surface);
  color: var(--primary-80);

  &:hover:not(:active, :disabled) {
    background-color: var(--secondary-30);
  }

  &:focus-visible {
    outline: 2px solid var(--border-col);
    outline-offset: 2px;
  }
}

.primary {
  --bg-col: var(--primary-80);

  border: none;
  background-color: var(--bg-col);
  color: var(--secondary-10);

  &:hover:not(:active, :disabled) {
    background-color: var(--primary-90);
  }

  &:focus-visible {
    outline: 2px solid var(--bg-col);
    outline-offset: 2px;
  }
}

.transparent {
  --border-col: transparent;
  outline: none;
  border: 2px solid var(--border-col);

  background-color: transparent;
  color: var(--primary-80);

  &:hover {
    &:not(:active, :disabled) {
      &:not(.icon) {
        text-shadow: 0 0 1px var(--primary-80);
      }

      &.icon {
        background-color: var(--secondary-30);
      }
    }
  }

  &:focus-visible {
    --border-col: var(--primary-80);
  }
}

.icon {
  padding: 0;
}

.primary,
.default {
  &:disabled {
    border-color: var(--gray-30);
    background-color: var(--gray-10);
  }
}
</style>
