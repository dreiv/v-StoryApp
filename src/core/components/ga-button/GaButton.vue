<script setup lang="ts">
import { computed, useCssModule, type ButtonHTMLAttributes } from 'vue'

export interface ButtonProps extends /* @vue-ignore */ ButtonHTMLAttributes {
  label?: string
  primary?: boolean
  ghost?: boolean
  transparent?: boolean
  icon?: boolean
}

const { primary, transparent, ghost, icon } = defineProps<ButtonProps>()

const style = useCssModule()
const classes = computed(() => {
  const classList = [style.button]

  if (primary) {
    classList.push(style.primary)
  } else if (ghost) {
    classList.push(style.ghost)
  } else if (transparent) {
    classList.push(style.transparent)
  } else {
    classList.push(style.secondary)
  }

  if (icon) {
    classList.push(style.icon)
  }

  return classList
})
</script>

<template>
  <button :class="classes">
    <slot>{{ label }}</slot>
  </button>
</template>

<style module>
.button {
  display: inline-flex;
  align-items: center;
  gap: var(--ga-size-spacing-03);

  border-radius: var(--ga-radius);
  padding-inline: var(--ga-size-spacing-05);
  min-width: var(--ga-size-spacing-08);
  height: var(--ga-size-spacing-08);

  font-weight: var(--ga-font-weight-medium);
  font-size: var(--ga-text-md-font-size);
  line-height: var(--ga-text-md-line-height);
  font-family: var(--ga-font-family-primary);
  letter-spacing: var(--ga-text-md-tracking);

  &:enabled {
    cursor: pointer;
  }

  &:focus-visible {
    outline: var(--ga-size-border-width-md) solid var(--ga-color-border-focus);
    outline-offset: var(--ga-size-spacing-01);
  }

  &:disabled {
    cursor: not-allowed;
  }
}

.primary {
  border: none;
  background-color: var(--ga-color-surface-action);
  color: var(--ga-color-text-on-action);

  &:not(:active, :disabled):is(:hover, :focus) {
    background-color: var(--ga-color-surface-action-hover);
  }

  &:disabled {
    background-color: var(--ga-color-surface-disabled);
    color: var(--ga-color-text-disable-selected);
  }
}

.secondary {
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-action);
  background-color: var(--ga-color-surface-primary);
  color: var(--ga-color-text-action);

  &:not(:active, :disabled):is(:hover, :focus) {
    background-color: var(--ga-color-border-action-hover-2);
  }

  &:disabled {
    border-color: var(--ga-color-border-disabled);
    background-color: var(--ga-color-surface-page);
    color: var(--ga-color-text-disabled);
  }
}

.ghost {
  border: none;
  background-color: transparent;
  color: var(--ga-color-text-action);

  &:not(:active, :disabled):is(:hover, :focus) {
    background-color: var(--ga-color-border-action-hover-2);
  }

  &:disabled {
    background-color: transparent;
    color: var(--ga-color-text-disabled);
  }
}

.transparent {
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-action);
  background-color: transparent;
  color: var(--ga-color-text-action);

  &:not(:active, :disabled):is(:hover, :focus) {
    background-color: var(--ga-color-border-action-hover-2);
  }

  &:disabled {
    border-color: var(--ga-color-border-disabled);
    background-color: transparent;
    color: var(--ga-color-text-disable-selected);
  }
}

.icon {
  justify-content: center;
  padding: 0;
}
</style>
