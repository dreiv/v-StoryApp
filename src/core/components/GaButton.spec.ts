import { shallowMount, type VueWrapper } from '@vue/test-utils'
import GaButton from './GaButton.vue'

describe('BaseButton', () => {
  let wrapper: VueWrapper
  beforeAll(() => {
    wrapper = shallowMount(GaButton, { slots: { default: 'mock content' } })
  })

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should apply the secondary class when no props are set', () => {
    expect(wrapper.find('.secondary').exists()).toBe(true)
  })

  it.each(['primary', 'ghost', 'transparent', 'icon'])(
    'should apply the correct class when the %s prop is set to true',
    async (prop) => {
      await wrapper.setProps({ [prop]: true })
      expect(wrapper.find(`.${prop}`).exists()).toBe(true)
    },
  )
})
