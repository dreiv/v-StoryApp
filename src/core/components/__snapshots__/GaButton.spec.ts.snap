// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseButton > should render correctly 1`] = `"<button class="base default">mock content</button>"`;
